src/dimensions.js

```diff
@@ -1,17 +1,94 @@
-import {extent} from "d3";
-import {projectionAspectRatio} from "./projection.js";
-import {isOrdinalScale} from "./scales.js";
+import {extent, max} from "d3";
+import {createProjection, projectionAspectRatio} from "./projection.js";
+import {createScales, isOrdinalScale} from "./scales.js";
 import {offset} from "./style.js";
+import {defaultWidth, monospaceWidth} from "./marks/text.js";
+import {createScaleFunctions, autoScaleRange, innerDimensions, outerDimensions} from "./scales.js";
 
-export function createDimensions(scales, marks, options = {}) {
+const marginMedium = 60;
+const marginLarge = 90;
+
+// When axes have "auto" margins, we might need to adjust the margins, after
+// seeing the actual tick labels. In that case we’ll compute the dimensions and
+// scales a second time.
+function autoMarginK(margin, {k: scale, labelAnchor, label}, options, mark, stateByMark, scales, dimensions, context) {
+  const {data, facets} = stateByMark.get(mark);
+  const {channels} = mark.initializer(data, facets, {}, scales, dimensions, context);
```

Comment:
Don’t you want to call mark.initialize here, not mark.initializer? The latter only applies to marks with an initializer, but here we want to initialize the mark (without assuming that it has an initializer).

Reply:
I think we need the initializer because the actual ticks depend on the instantiated scales. We know that the mark has an initializer because it can only be a (tick label) axis mark.

Reply:
My point is that we should follow the same code path that Plot.plot follows:

- call _mark_.initialize first
- if _mark_.initializer exists, call that second

We shouldn’t make such a strong assumption about how the mark is implemented. Also, I wasn’t expecting that the mark would already be initialized (using stateByMark); I was thinking we would initialize it ourselves here. So that further complicates matters.

Reply:
When we call autoMarginK all the marks have been initialized already, which is why we can avoid the cost of re-initializing them. I can see how reaching for stateByMark can feel messy, though. 🥺

Reply:
If the mark is already initialized here, why do we need to invoke its initializer again?